feat: allow disk inject to delete files

glattercj opened the pull request:
Adds the ability to delete files while doing disk inject via a keyword as the src. E.g.:

`disk inject window7_miniccc.qc2 files "-DELETE-":"Program Files/miniccc.exe"`

jacdavi reviewed (approved):
Tested and looks good. Two UX considerations though:

* Would it make sense to have the syntax be `disk inject <image> delete files <files like /path/to/dst>`? I realize it would require a bit more refactoring, but on the surface it seems more intuitive to me
* Would it be useful to log a warning if a file the user tries to delete doesn't exist? I fat-fingered a path in testing, and wouldn't have noticed if I wasn't manually verifying the image after. `os.RemoveAll` just returns `nil` in that case, so we'd have to separately check if the file exists.

glattercj commented:
@jacdavi 

- Updated to use your suggested format. E.g.:

    ```bash
    disk inject window7_miniccc.qc2 delete files "Program Files/miniccc.exe"
    ```

- Added warning if path doesn't exist for deletion

jacdavi commented:
Thanks I think those are both really good changes! It does seem like there's an issue parsing when deleting a single file. Based on the docstring I think this should work?

```
minimega[minimega]$ disk inject /scratch/images/test.qcow2 delete files /etc/setup.sh
Error (ccc19.40g): malformed command; expected src:dst pairs
```

glattercj commented:
> Thanks I think those are both really good changes! It does seem like there's an issue parsing when deleting a single file. Based on the docstring I think this should work?
> 
> ```
> minimega[minimega]$ disk inject /scratch/images/test.qcow2 delete files /etc/setup.sh
> Error (ccc19.40g): malformed command; expected src:dst pairs
> ```

@jacdavi Good catch, I missed the single file case 🤦 . Updated and tested.

jacdavi commented on cmd/minimega/disk.go:
```diff
@@ -343,6 +355,37 @@ func parseInjectPairs(files []string) (map[string]string, error) {
+			pairs, err = parseInjectPairs(v)
+		}
+	case string:
+		if strings.Contains(v, ",") {
```
Can the single case also be added here. Otherwise it doesn' t work with options specified

```
minimega[minimega]$ disk inject /scratch/images/test.qcow2 options "-o offset=0" delete files /bin/foo
Error (ccc19.40g): malformed command; expected src:dst pairs
```

glattercj replied:
Fixed

jacdavi commented on cmd/minimega/disk.go:
```diff
@@ -384,15 +427,24 @@ func cliDisk(ns *Namespace, c *minicli.Command, resp *minicli.Response) error {
 
-		pairs, err := parseInjectPairs(c.ListArgs["files"])
+		var files interface{}
+		if len(options) == 0 {
```
I accidentally found a panic here if you use `disk inject image.qcow2 options "" files /bin/foo` as this if evaluates to true, but the parser has made `files` a string. 

Fix should just be `if _, ok := c.StringArgs["options"]; !ok {` so we just check whether the user specified options

glattercj replied:
I'm not able to repro this. I changed the `if` check to match your fix but here's what I'm getting both before and after making the change:

```bash
$ disk inject image.qcow2 options "" files /bin/foo
Error (crn1): invalid command: `disk inject image.qcow2 options  files /bin/foo`
```

If I add the `delete` keyword it just returns without any output. Looking at some debug output the minicli.Command just turns the empty quotes into a blank so `c.StringArgs["options"]` returns the next token being `delete`.

```bash
$ disk inject image.qcow2 options "" delete files /bin/foo

```

jacdavi replied:
Hmm it was happening to me every time. Regardless it looks good now

```
minimega[minimega]$ disk inject /scratch/images/test.qcow2 options "" files /bin/foo
panic: runtime error: index out of range [0] with length 0

goroutine 99 [running]:
main.parseFiles({0x8d0cc0?, 0xc0000a4df0?})
        /scratch/minimega/cmd/minimega/disk.go:370 +0x276
main.cliDisk(0x0?, 0xc0000c2780, 0xc00053e400)
        /scratch/minimega/cmd/minimega/disk.go:442 +0x5eb
main.wrapSimpleCLI.func1(0xc0000c2780?, 0xc0001129c0?)
        /scratch/minimega/cmd/minimega/cli.go:104 +0x8b
github.com/sandia-minimega/minimega/v2/pkg/minicli.ProcessCommand.func1()
        /scratch/minimega/pkg/minicli/minicli.go:173 +0x196
created by github.com/sandia-minimega/minimega/v2/pkg/minicli.ProcessCommand
        /scratch/minimega/pkg/minicli/minicli.go:160 +0xca
```

jacdavi reviewed (changes requested):
Sorry for so much back and forth. Two final things and then it should be all set!

jacdavi commented on cmd/minimega/disk.go:
```diff
@@ -328,6 +355,39 @@ func parseInjectPairs(files []string) (map[string]string, error) {
+	var err error
+	switch v := files.(type) {
+	case []string:
+		if !(sliceContainsCharacters(v, ",:")) {
```
Final final thing I promise :) 

With this change we no longer get the error "malformed command; expected src:dst pairs" when running `disk inject /scratch/images/test.qcow2 files /bin/foo`.

Maybe the code should just be explicit whether it expects single files or pairs based on the delete flag?

glattercj replied:
@jacdavi I changed the logic a bit to handle this case and tried all the previous ones you had asked about. I think it covers them all. Let me know what you think.

glattercj replied:
@jacdavi bump

jacdavi replied:
Thanks for the reminder, looks good now!
